xarray/core/dataset.py

```diff
@@ -369,24 +400,24 @@ def __init__(self, data_vars=None, coords=None, attrs=None,
         else:
             compat = 'broadcast_equals'
 
-        self._variables = OrderedDict()  # type: OrderedDict[Any, Variable]
-        self._coord_names = set()
-        self._dims = {}  # type: Dict[Any, int]
+        self._variables = \
+            OrderedDict()  # type: OrderedDict[Hashable, Variable]
```

Comment:
I think anything used as a dict key already has to be `Hashable`, which is why I used `Any` previously. But maybe type checkers handle these differently?

Reply:
I think they'll be equivalent. Which do you think we should use?

Reply:
I like `Any` just because it's shorter. But not a strong preference.

Reply:
It is true that a key of an ordereddict is always hashable. However, mypy doesn't know it, and the moment you pull it out of the dict and you pass it to some helper function it's nice to track what it is. Also, a Mapping does not require its keys to be hashable, but when xarray accepts a Mapping input (such as what is accepted by DataArray.attrs.setter, DataArray.coords.setter, etc.)  it should always state Mapping[Hashable, ...] since it's going to be used to feed an OrderedDict.